File: C++/Longest subsequence-1.cpp
class Solution {
  public:
    int longestSubseq(int n, vector<int> &a) {
        int k = *max_element(a.begin(),a.end());
        int maxi=-1;
        vector<int>dp(k+2,0);
        for(int x:a){
        dp[x] = max(dp[x-1],dp[x+1])+1; 
        maxi = max(maxi,dp[x]);
        }
        return maxi;
    }
};

File: JAVA/Longest subsequence-1.java
class Solution {
    public static int longestSubseq(int n, int[] a) {
        return solve(a,0,n,-1);
    }
    public static int solve(int arr[],int idx ,int n,int prev){
        if(idx >= n){
            return 0;
        }
        int max = 0;
        for(int x = idx;x<n;x++){
            int diff = Math.abs(prev-arr[x]);
            if(diff == 1 || prev == -1){
                max = Math.max(max,1+solve(arr,x+1,n,arr[x]));
            }
            
        }
        return max;
    }
}
